From eb3642a63a4b76721d1d3857654137d69aee122e Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Tue, 6 Dec 2016 18:53:00 +0000
Subject: [PATCH] OS-5741 incorrect 'free -m' when over memory cap Reviewed by:
 Patrick Mooney <patrick.mooney@joyent.com> Approved by: Patrick Mooney
 <patrick.mooney@joyent.com>

---
 usr/src/uts/common/brand/lx/procfs/lx_prvnops.c  |  2 ++
 usr/src/uts/common/brand/lx/syscall/lx_sysinfo.c | 11 +++++++++--
 usr/src/uts/common/fs/lxproc/lxpr_vnops.c        |  2 ++
 3 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c b/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
index fb06882064..edfada483c 100644
--- a/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
+++ b/usr/src/uts/common/brand/lx/procfs/lx_prvnops.c
@@ -3624,6 +3624,8 @@ lxpr_read_meminfo(lxpr_node_t *lxpnp, lxpr_uiobuf_t *uiobuf)
 	} else {
 		total_mem = zone->zone_phys_mem_ctl;
 		free_mem = zone->zone_phys_mem_ctl - zone->zone_phys_mem;
+		if (free_mem < 0)
+			free_mem = 0;
 	}
 
 	if (global || zone->zone_max_swap_ctl == UINT64_MAX) {
diff --git a/usr/src/uts/common/brand/lx/syscall/lx_sysinfo.c b/usr/src/uts/common/brand/lx/syscall/lx_sysinfo.c
index 449d5882d4..387471c0f5 100644
--- a/usr/src/uts/common/brand/lx/syscall/lx_sysinfo.c
+++ b/usr/src/uts/common/brand/lx/syscall/lx_sysinfo.c
@@ -21,7 +21,7 @@
 /*
  * Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
  * Use is subject to license terms.
- * Copyright 2015 Joyent, Inc.
+ * Copyright 2016 Joyent, Inc.
  */
 
 #include <vm/anon.h>
@@ -100,8 +100,15 @@ lx_sysinfo_common(lx_sysinfo_t *si)
 		zphysmem = physmem;
 		zfreemem = freemem;
 	} else {
+		int64_t freemem;
+
 		zphysmem = btop(zone->zone_phys_mem_ctl);
-		zfreemem = btop(zone->zone_phys_mem_ctl - zone->zone_phys_mem);
+		freemem = zone->zone_phys_mem_ctl - zone->zone_phys_mem;
+		if (freemem > 0) {
+			zfreemem = btop(freemem);
+		} else {
+			zfreemem = 0;
+		}
 	}
 
 	if (zone->zone_max_swap_ctl == UINT64_MAX) {
diff --git a/usr/src/uts/common/fs/lxproc/lxpr_vnops.c b/usr/src/uts/common/fs/lxproc/lxpr_vnops.c
index 9c996891f3..9b02fea244 100644
--- a/usr/src/uts/common/fs/lxproc/lxpr_vnops.c
+++ b/usr/src/uts/common/fs/lxproc/lxpr_vnops.c
@@ -1454,6 +1454,8 @@ lxpr_read_meminfo(lxpr_node_t *lxpnp, lxpr_uiobuf_t *uiobuf)
 	} else {
 		total_mem = zone->zone_phys_mem_ctl;
 		free_mem = zone->zone_phys_mem_ctl - zone->zone_phys_mem;
+		if (free_mem < 0)
+			free_mem = 0;
 	}
 
 	if (global || zone->zone_max_swap_ctl == UINT64_MAX) {
-- 
2.21.0

